DROD/RoomScreen.cpp

```diff
@@ -427,13 +428,73 @@ void CRoomScreen::AddNoticesDialog()
 void CRoomScreen::OpenNoticesBox(CRoomWidget* pRoomWidget)
 {
 	this->pNoticesDialog->SetBetweenEventsHandler(GetEventHandlerWidget()); //keep updating room effects
-	const UINT returnTag = this->pNoticesDialog->Display(false);
+	UINT returnTag = 0;
+	do {
+		if (returnTag == TAG_NOTICESDELETE) {
+			UINT selected = this->pNoticesList->GetSelectedItem();
+			this->pNoticesList->RemoveItem(selected);
+			this->notices.erase(this->notices.find(selected));
+		}
+		returnTag = this->pNoticesDialog->Display(false);
+		this->UpdateNoticesButton();
+
+	} while (returnTag == TAG_NOTICESDELETE);
+
+	switch (returnTag) {
+	case TAG_NOTICESNOTICE:
+		// This button should go to whatever the notice was about. 
+		UINT selected = this->pNoticesList->GetSelectedItem();
+		std::map<UINT, CNetNotice>::const_iterator item = this->notices.find(selected);
+		if (item != this->notices.end()) {
+			// Found it!
+			switch (item->second.type) {
+			case NOTICE_NEWHOLD:
+			case NOTICE_UPDATEDHOLD:
+				// go to the hold select screen so the user can update it
+				g_pTheNet->SetDownloadHold(item->second.dwServerHoldId);
+				GoToScreen(SCR_HoldSelect);
+				break;
+
+			case NOTICE_GENERICURL:
+				// open a web browser to the specified URL
+				OpenExtBrowser(item->second.url.c_str());
+				break;
+
+			case NOTICE_ROOMSPECIFIC:
+				// Go to the hold/level/room specified.  
+				break;
+			}
+		}
+		break;
+	}
+
 	this->pNoticesDialog->SetBetweenEventsHandler(NULL);
 
 	pRoomWidget->DirtyRoom();
 	pRoomWidget->Paint();
 }
 
+//*****************************************************************************
+const void CRoomScreen::UpdateNoticesButton()
+{
+	const int num = this->notices.size();
+	CButtonWidget* pButton = dynamic_cast<CButtonWidget*>(this->GetWidget(TAG_OPEN_NOTICES));
+	string label;
+	WSTRING wLabel;
+	if (num >= 10) {
+		label = "...";
+	}
+	else {
+		char buffer[5];
```

Comment:
I know we're just refactoring code location, but while we're at it, let's set buffer's size to 12 to avoid potential overflow issues (even if not realistic, we don't want to open the door to them).

Reply:
This is in an else. The if condition is "if (num >= 10)".  Five is safe here.